### ISSUE TITLE
Question about no-parameter sharing behavior in `MultiAgentMLP`

### ISSUE THREAD

**Issue by khinggan** [pytorch/rl#2904]
Hi,

First of all, thank you for the amazing work on TorchRL — the library is super helpful!

I'm currently working through the [Multi-Agent Reinforcement Learning (PPO) with TorchRL Tutorial](https://pytorch.org/rl/stable/tutorials/multiagent_ppo.html). By default, the example uses a shared policy network with `share_parameters_policy = True`. I’d like to try the no-parameter sharing (no-PS) version and encountered some confusion when setting that up.

Here's a minimal example I wrote:
```python
from torchrl.modules.models.multiagent import MultiAgentMLP
mlp = MultiAgentMLP(
    n_agent_inputs=18,
    n_agent_outputs=2,
    n_agents=6,
    centralized=False,
    share_params=False,
    depth=2,
)
print(mlp)
print(mlp.params)
```

Since I’ve set `share_params=False` and `n_agents=6`, I expected the model to include a list of independent MLPs, something like:
```python
MultiAgentMLP(
  (agent_networks): ModuleList(
    (0-5): 6 x MLP(
      (0): Linear(in_features=3, out_features=32, bias=True)
      (1): Tanh()
      (2): Linear(in_features=32, out_features=32, bias=True)
      (3): Tanh()
      (4): Linear(in_features=32, out_features=2, bias=True)
    )
  )
)
```
However, print(mlp) gives me this:
```python
MultiAgentMLP(
    MLP(
      (0): Linear(in_features=18, out_features=32, bias=True)
      (1): Tanh()
      (2): Linear(in_features=32, out_features=32, bias=True)
      (3): Tanh()
      (4): Linear(in_features=32, out_features=2, bias=True)
    ),
    n_agents=6,
    share_params=False,
    centralized=False,
    agent_dim=-2)
```
It seems like there's still only one MLP network, not six. But the params show extended shapes that seem agent-specific:
```python
TensorDictParams(params=TensorDict(
    fields={
        0: TensorDict(
            fields={
                bias: Parameter(shape=torch.Size([6, 32]), device=cpu, dtype=torch.float32, is_shared=False),
                weight: Parameter(shape=torch.Size([6, 32, 18]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([6]),
            device=None,
            is_shared=False),
        2: TensorDict(
            fields={
                bias: Parameter(shape=torch.Size([6, 32]), device=cpu, dtype=torch.float32, is_shared=False),
                weight: Parameter(shape=torch.Size([6, 32, 32]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([6]),
            device=None,
            is_shared=False),
        4: TensorDict(
            fields={
                bias: Parameter(shape=torch.Size([6, 2]), device=cpu, dtype=torch.float32, is_shared=False),
                weight: Parameter(shape=torch.Size([6, 2, 32]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([6]),
            device=None,
            is_shared=False)},
    batch_size=torch.Size([6]),
    device=None,
    is_shared=False))
```

So I’m a bit confused and would love clarification on two points:

1. For `share_params=False`, should `MultiAgentMLP` internally maintain a list of MLPs (one per agent)? Or is this parameter expansion via `TensorDictParams` the intended behavior?

2. How does MultiAgentMLP ensure that each agent’s parameters are updated independently during training (e.g., when computing gradients and applying optimizer steps)? 

Thanks a lot for your time, and again, great work on TorchRL!

Best,

**Comment by matteobettini** [comment]
The multi-agent mlp with no parameter sharing uses a single skeleton network and different parameters for each agent. It then uses vmap to do the forward pass. This is because the network structure is the same for all agents and there is no need with replicating it.

The result is identical to having different networks bit much faster. In fact, the previous implementation had different networks and we checked that the results are identical

**Comment by khinggan** [comment]
@matteobettini Thank you for the response, I got it and closing the issue now.
